python/pylibcudf/pylibcudf/column.pyx

```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+    # - index  = [0, 1,    2, 3, 4,    5, 6,    7, 8, 9]
+    #
+    # Row values:
+    # - row 0 -> valid -> 1
+    # - row 1 -> null  -> 0
+    # - row 2 -> valid -> 1
+    # - row 3 -> valid -> 1
+    # - row 4 -> null  -> 0
+    # - row 5 -> valid -> 1
+    # - row 6 -> null  -> 0
+    # - row 7 -> valid -> 1
+    # - row 8 -> valid -> 1
+    # - row 9 -> null  -> 0
+    #
+    # Build the bitmap:
+    # - index     0,1,2,3,4,5,6,7       76543210 (row_index -> bit_index)
+    # - Byte 0 = [1,0,1,1,0,1,0,1] -> 0b10101101
+    # - Byte 1 = [1,0,0,0,0,0,0,0] -> 0b00000001
+    #
+    # Check the 7th element:
+    # row_index = 6
+    # - row 5 -> valid -> 1
+    # - row 6 -> null  -> 0
+    # - row 7 -> valid -> 1
+    # - row 8 -> valid -> 1
+    # - row 9 -> null  -> 0
+    #
+    # Build the bitmap:
+    # - index     0,1,2,3,4,5,6,7       76543210 (row_index -> bit_index)
+    # - Byte 0 = [1,0,1,1,0,1,0,1] -> 0b10101101
+    # - Byte 1 = [1,0,0,0,0,0,0,0] -> 0b00000001
+    #
+    # Check the 7th element:
+    # row_index = 6
+    # offset = 0
+    # bit_index = offset + row_index = 6
+    # 1. Choose the byte
+    #    bit_index >> 3 = 0
+    #    null_bitmap[0] = 0b10101101
+    # 2. Get the bit position within that byte (bit_index mod 8)
+    #    bit_index & 7 = 6 & 7 = 6
+    # 3. Shift the target bit to the least significant bit position
+    #    0b10101101 >> 6 = 0b00000010
+    # 4. Remove all other bits
+    #    0b00000010 & 1 = 0 (=> the 7th element is null)
+    if null_bitmap == NULL:
+        return 1
+    return ((null_bitmap[bit_index >> 3] >> (bit_index & 7)) & 1) != 0
+
+
+cdef inline void _set_item(list out, int64_t i, object obj) except *:
+    # https://docs.python.org/3/c-api/list.html#c.PyList_SetItem
+    Py_INCREF(obj)
+    PyList_SetItem(out, i, obj)
+
+
+cdef list _arrow_to_pylist_impl(type_id dtype, ArrowArray* arr):
+    cdef size_t n = <size_t>arr.length
+    cdef list out = <list>PyList_New(n)
+    if out is None:
+        raise MemoryError("Unable to create new python list")
+
+    if n == 0:
+        return out
+
+    cdef const void** buffers = arr.buffers
+    cdef int64_t offset = arr.offset
+    cdef size_t bit_index
+    cdef int64_t i, N = n
+    cdef const uint8_t* null_bm = NULL
+    cdef const char* numeric_base = NULL
+    cdef size_t item_size = 0
+    cdef const uint8_t* bool_base = NULL
+    cdef const int32_t* str_offsets = NULL
+    cdef const char* str_base = NULL
+    cdef bint is_bool = dtype == type_id.BOOL8
+    cdef bint is_string = dtype == type_id.STRING
+    cdef bint check_valid = arr.null_count != 0
+
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
+
+    if dtype in (
+        type_id.INT8, type_id.INT16, type_id.INT32, type_id.INT64,
+        type_id.UINT8, type_id.UINT16, type_id.UINT32, type_id.UINT64,
+        type_id.FLOAT32, type_id.FLOAT64, type_id.BOOL8, type_id.STRING
+    ):
```

Comment:
I would invert the condition and throw the exception. That way the rest of the function can be deindented. Also for this many values it might be more efficient to use a set than a tuple.